python-avd/pyavd/_eos_designs/structured_config/underlay/dhcp_server.py

```diff
@@ -0,0 +1,114 @@
+
+from pyavd._utils import get
+
+from .utils import UtilsMixin
+
+
+class DhcpServerMixin(UtilsMixin):
+    """
+    Mixin Class used to generate structured config for one key.
+
+    Class should only be used as Mixin to a AvdStructuredConfig class.
+    """
+
+    _hostvars: dict
+
+    @cached_property
+    def _subnets(self) -> list:
+        """
+        Returns a list of dhcp subnets for downstream p2p interfaces.
+
+        Used for l3 inband ztp/ztr.
+        """
+        subnets = []
+        for peer in self._avd_peers:
+            peer_facts = self.shared_utils.get_peer_facts(peer, required=True)
+            for uplink in peer_facts["uplinks"]:
+                if (
+                    uplink["peer"] == self.shared_utils.hostname
+                    and uplink["type"] == "underlay_p2p"
+                    and uplink.get("ip_address")
+                    and "unnumbered" not in uplink["ip_address"]
+                    and get(peer_facts, "inband_ztp")
+                ):
+                    subnet = {
+                        "subnet": str(ip_network(f"{uplink['peer_ip_address']}/{uplink['prefix_length']}", strict=False)),
+                        "ranges": [{"start": str(uplink["ip_address"]), "end": str(uplink["ip_address"])}],
+                        "name": f"inband ztp for {peer}-{uplink['interface']}",
+                        "default_gateway": f"{uplink['peer_ip_address'].split('/')[0]}",
+                    }
+                    subnets.append(subnet)
+
+        return subnets
+
+    @cached_property
+    def _ipv4_ztp_boot_file(self) -> str | None:
+        """Returns the file name to allow for ZTP to CV."""
+        custom_bootfile = get(self._hostvars, "inband_ztp_bootstrap_file")
+        if custom_bootfile:
+            return custom_bootfile
+        cvp_instance_ips = get(self._hostvars, "cvp_instance_ips")
+        if not cvp_instance_ips:
+            return None
+        if "arista.io" in cvp_instance_ips[0]:
+            return "https://www.arista.io/ztp/bootstrap"
+
+        return f"https://{cvp_instance_ips[0]}/ztp/bootstrap"
+
+    @cached_property
+    def _dns_servers(self) -> list | None:
+        """Returns the list of name servers."""
+        dns_servers = get(self._hostvars, "name_servers")
+        if not dns_servers:
+            return None
+
+        return dns_servers
+
+    @cached_property
+    def _ntp_servers(self) -> list | None:
+        """Returns the list of name servers."""
+        ntp_servers_settings = get(self._hostvars, "ntp_settings.servers")
+        if not ntp_servers_settings:
+            return None
+
+        ntp_servers = []
+        for ntp_server in ntp_servers_settings:
+            # Validate NTP server IP address
+            try:
+                ntp_server_ip = IPv4Address(ntp_server["name"])
+            except AddressValueError:
+                continue
+            ntp_servers.append(str(ntp_server_ip))
```

Comment:
We should not force someone to use IP as NTP server. DNS names are fine, so please remove this check. We should make sure to also give the dns options in dhcp in this feature.

Reply:
EOS doesn't accept an fqdn for the NTP server value.  That's why there is the check.

Reply:
I would prefer not to resolve stuff during build, since we would no longer be idempotent. It sounds weird to me that we cannot use dns name here... I will unicast you.

Reply:
So after discussion we settle on ignoring NTP servers with names, and then only error out if we have no NTP servers left.